src/sas/sasgui/perspectives/fitting/gpu_options.py

```diff
@@ -268,12 +269,17 @@ def on_test(self, event):
 
         try:
             env = sasmodels.kernelcl.environment()
-            clinfo = [(ctx.devices[0].platform.vendor,
-                       ctx.devices[0].platform.version,
-                       ctx.devices[0].vendor,
-                       ctx.devices[0].name,
-                       ctx.devices[0].version)
-                      for ctx in env.context]
+            dtype64 = np.dtype('float64')
+            dtype32 = np.dtype('float32')
+            if dtype64 in env.context:
+                ctx = env.context[dtype64].devices[0]
+            else:
+                ctx = env.context[dtype32].devices[0]
+            clinfo = (ctx.platform.vendor,
+                        ctx.platform.version,
+                        ctx.vendor,
+                        ctx.name,
+                        ctx.version)
```

Comment:
(1) There may be two contexts active simultaneously, such as the GPU for single precision calculations and the CPU for double precision calculations.  The models which require double precision will automatically choose the double precision context.  So really, we ought to report both contexts unless they happen to be identical.

(2) Looking at sasmodels/kernelcl.py, F64 will always be in the context, though it may be None.  This will cause line 278... to fail.

(3) You can use `from sasmodels.generate import F32, F64` so that you know that the keys will be in the correct form for env.context.

(4) Trivial point: lint will complain that you are indented two characters too deep.

(5) Not sure how this code relates to the rest of the OpenCL context selection GUI.  That is, if you select a F32-only device for your calculations, does that mean the intel OpenCL driver for the CPU will not be used for double precision models?

Reply:
If I understood comments correctly 1-4  should be fixed with the latest commit. I am not sure if (5) applies here. It seems to me that this shouldn't affect actual driver selection (it is only pulling out information)